File: src/sys/dotnet/fan/sys/RegexMatcher.cs
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   27 Dec 07  Andy Frank  Creation
//

using System.Text.RegularExpressions;

namespace Fan.Sys
{
  /// <summary>
  /// RegexMatcher
  /// </summary>
  public sealed class RegexMatcher : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructors
  //////////////////////////////////////////////////////////////////////////

    internal RegexMatcher(Match match, string source)
    {
      this.m_match = match;
      this.m_source = source;
    }

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public override Type @typeof() { return Sys.RegexMatcherType; }

  //////////////////////////////////////////////////////////////////////////
  // Methods
  //////////////////////////////////////////////////////////////////////////

    public bool matches()
    {
      // to match java
      return (m_match.Success) ? m_source.Length == m_match.Length : false;
    }

    public bool find()
    {
      return true; //matcher.find());
    }

    public long groupCount()
    {
      // to match java
      return m_match.Groups.Count-1;
    }

    public string group() { return group(0); }
    public string group(long group)
    {
      // to match java
      if (!matches()) throw new System.Exception();
      if (group < 0 || group >= m_match.Groups.Count)
        throw IndexErr.make(group).val;

      return m_match.Groups[(int)group].Value;
    }

    public long start() { return start(0); }
    public long start(long group)
    {
      // to match java
      if (!matches()) throw new System.Exception();
      if (group < 0 || group >= m_match.Groups.Count)
        throw IndexErr.make(group).val;

      return m_match.Groups[(int)group].Index;
    }

    public long end() { return end(0); }
    public long end(long group)
    {
      // to match java
      if (!matches()) throw new System.Exception();
      if (group < 0 || group >= m_match.Groups.Count)
        throw IndexErr.make(group).val;

      Group g = m_match.Groups[(int)group];
      return g.Index + g.Length;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    Match m_match;
    string m_source;

  }
}

File: src/sys/java/fan/sys/RegexMatcher.java
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   26 Dec 07  Brian Frank  Creation
//
package fan.sys;

import java.util.regex.*;

/**
 * RegexMatcher is used in matching operations on a regular expression.
 */
public final class RegexMatcher
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Constructors
//////////////////////////////////////////////////////////////////////////

  RegexMatcher(Matcher matcher)
  {
    this.matcher = matcher;
  }

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public Type typeof() { return Sys.RegexMatcherType; }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public final boolean matches()
  {
    return matcher.matches();
  }

  public final boolean find()
  {
    return matcher.find();
  }

  public final String replaceFirst(String replacement)
  {
    return matcher.replaceFirst(replacement);
  }

  public final String replaceAll(String replacement)
  {
    return matcher.replaceAll(replacement);
  }

  public final long groupCount()
  {
    return matcher.groupCount();
  }

  public final String group() { return group(0L); }
  public final String group(long group)
  {
    try
    {
      return matcher.group((int)group);
    }
    catch (IllegalStateException e)
    {
      throw Err.make(e.getMessage());
    }
    catch (IndexOutOfBoundsException e)
    {
      throw IndexErr.make(group);
    }
  }

  public final long start() { return start(0L); }
  public final long start(long group)
  {
    try
    {
      return matcher.start((int)group);
    }
    catch (IllegalStateException e)
    {
      throw Err.make(e.getMessage());
    }
    catch (IndexOutOfBoundsException e)
    {
      throw IndexErr.make(group);
    }
  }

  public final long end() { return end(0L); }
  public final long end(long group)
  {
    try
    {
      return Long.valueOf(matcher.end((int)group));
    }
    catch (IllegalStateException e)
    {
      throw Err.make(e.getMessage());
    }
    catch (IndexOutOfBoundsException e)
    {
      throw IndexErr.make(group);
    }
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  Matcher matcher;
}

